api/bucketsd/service.go

```diff
@@ -1497,6 +1518,367 @@ func (s *Service) PushPath(server pb.APIService_PushPathServer) (err error) {
+	doneCh chan<- addedFile,
+	errCh chan<- error,
+) (*fileAdder, error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	fa, ok := q.q[pth]
+	if ok {
+		return fa, nil
+	}
+
+	key, err := addFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	reader, writer := io.Pipe()
+	fa = &fileAdder{
+		reader: reader,
+		writer: writer,
+	}
+	q.q[pth] = fa
+
+	eventCh := make(chan interface{})
+	chSize := make(chan string)
+	go func() {
+		for e := range eventCh {
+			event, ok := e.(*iface.AddEvent)
+			if !ok {
+				log.Error("unexpected event type")
+				continue
+			}
+			if event.Path != nil {
+				chSize <- event.Size // Save size for use in the final response
+			}
+		}
+	}()
+
+	var r io.Reader
+	if key != nil {
+		r, err = dcrypto.NewEncrypter(reader, key)
+		if err != nil {
+			return nil, fmt.Errorf("creating decrypter: %v", err)
+		}
+	} else {
+		r = reader
+	}
+
+	go func() {
+		defer close(eventCh)
+		res, err := ufs.Add(
+			ctx,
+			ipfsfiles.NewReaderFile(r),
+			options.Unixfs.CidVersion(1),
+			options.Unixfs.Pin(false),
+			options.Unixfs.Progress(true),
+			options.Unixfs.Events(eventCh),
+		)
+		if err != nil {
+			errCh <- fmt.Errorf("adding file: %v", err)
+			return
+		}
+		size := <-chSize
+		added, err := strconv.Atoi(size)
+		if err != nil {
+			errCh <- fmt.Errorf("getting file size: %v", err)
+			return
+		}
+		doneCh <- addedFile{path: pth, resolved: res, size: int64(added)}
+	}()
+
+	return fa, nil
+}
+
+func (s *Service) PushPaths(server pb.APIService_PushPathsServer) error {
+	log.Debugf("received push paths request")
+
+	ctx, cancel := context.WithCancel(server.Context())
+	defer cancel()
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return errDBRequired
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	req, err := server.Recv()
+	if err != nil {
+		return fmt.Errorf("on receive: %v", err)
+	}
+	var buckKey, buckRoot string
+	switch payload := req.Payload.(type) {
+	case *pb.PushPathsRequest_Header_:
+		buckKey = payload.Header.Key
+		buckRoot = payload.Header.Root
+	default:
+		return fmt.Errorf("push bucket path header is required")
+	}
+
+	lck := s.Semaphores.Get(buckLock(buckKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck := &tdb.Bucket{}
+	err = s.Buckets.GetSafe(ctx, dbID, buckKey, buck, tdb.WithToken(dbToken))
+	if err != nil {
+		return fmt.Errorf("getting bucket: %v", err)
+	}
+	if buckRoot != "" && buckRoot != buck.Path {
+		return status.Error(codes.FailedPrecondition, buckets.ErrNonFastForward.Error())
+	}
+
+	var wg sync.WaitGroup
+	addedCh := make(chan addedFile)
+	doneCh := make(chan struct{})
+	errCh := make(chan error)
+	go func() {
+		queue := newFileQueue()
+		for {
+			req, err := server.Recv()
+			if err == io.EOF {
+				wg.Wait() // Request ended normally, wait for pending jobs
+				close(doneCh)
+				return
+			} else if err != nil {
+				errCh <- fmt.Errorf("on receive: %v", err)
+				return
+			}
+			switch payload := req.Payload.(type) {
+			case *pb.PushPathsRequest_Chunk_:
```

Comment:
Protoc spits those out when you use the `oneof payload` directive:
```
oneof payload {
    Header header = 1;
    Chunk chunk = 2;
}
```

Reply:
@sanderpick , weird since `master` also uses `oneof` and doesn't seem to have generated that?
- [`master` proto ](https://github.com/textileio/textile/blob/a55836b9aa0490a146432b36b3c0b23f1f9b7006/api/bucketsd/pb/bucketsd.proto#L95)
- [`master` same line of code without _](https://github.com/textileio/textile/blob/a55836b9aa0490a146432b36b3c0b23f1f9b7006/api/bucketsd/service.go#L1410).

Maybe some gRPC update that changed something else.

Reply:
`chunk` changed from `bytes` to an actual type. You can [see the underscore](https://github.com/textileio/textile/blob/master/api/bucketsd/service.go#L1347) in master used to switch on `header` which is also a type in master.

Reply:
Oh, I see! :)